Mathlib/CategoryTheory/Closed/FunctorCategory/Basic.lean

```diff
@@ -0,0 +1,148 @@
+we obtain that the category `J ⥤ C` is monoidal closed if `C` has suitable
+limits.
+
+-/
+
+universe v₁ v₂ u₁ u₂
+
+namespace CategoryTheory
+
+open Category Limits MonoidalCategory
+
+namespace MonoidalClosed
+
+namespace FunctorCategory
+
+open Enriched.FunctorCategory
+
+variable {C : Type u₁} [Category.{v₁} C] [MonoidalCategory C] [MonoidalClosed C]
+  {J : Type u₂} [Category.{v₂} J]
+  [∀ (F₁ F₂ : J ⥤ C), HasFunctorEnrichedHom C F₁ F₂]
+
+attribute [local simp] enrichedCategorySelf_hom enrichedCategorySelf_id
+  enrichedCategorySelf_comp enrichedOrdinaryCategorySelf_eHomWhiskerLeft
+  enrichedOrdinaryCategorySelf_eHomWhiskerRight
+
+section
+
+variable {F₁ F₂ F₂' F₃ F₃' : J ⥤ C}
+
+/-- The bijection `(F₁ ⊗ F₂ ⟶ F₃) ≃ (F₂ ⟶ functorEnrichedHom C F₁ F₃)` when `F₁`, `F₂`
+and `F₃` are functors `J ⥤ C`, and `C` is monoidal closed. -/
+noncomputable def homEquiv : (F₁ ⊗ F₂ ⟶ F₃) ≃ (F₂ ⟶ functorEnrichedHom C F₁ F₃) where
+  toFun f :=
+    { app j := end_.lift (fun k ↦ F₂.map k.hom ≫ curry (f.app k.right))
+        (fun k₁ k₂ φ ↦ by
+          dsimp
+          simp only [enrichedOrdinaryCategorySelf_eHomWhiskerLeft, Category.assoc,
+            enrichedOrdinaryCategorySelf_eHomWhiskerRight]
+          rw [← curry_natural_left_assoc, ← curry_natural_left_assoc,
+            ← curry_natural_right, curry_pre_app, Category.assoc,
+            ← f.naturality φ.right, Monoidal.tensorObj_map, tensorHom_def_assoc,
+            ← Under.w φ, Functor.map_comp, MonoidalCategory.whiskerLeft_comp_assoc,
+            whisker_exchange_assoc]) }
+  invFun g :=
+    { app j := uncurry (g.app j ≫ enrichedHomπ C _ _ (Under.mk (𝟙 j)) )
+      naturality j j' φ := by
+        dsimp
+        rw [← uncurry_natural_right, tensorHom_def'_assoc, ← uncurry_pre_app,
+          ← uncurry_natural_left, Category.assoc, Category.assoc,
+          NatTrans.naturality_assoc, functorEnrichedHom_map,
+          end_.lift_π_assoc, enrichedOrdinaryCategorySelf_eHomWhiskerRight]
+        dsimp
+        rw [pre_id, NatTrans.id_app, enrichedOrdinaryCategorySelf_eHomWhiskerLeft,
+          Functor.map_id, Category.comp_id, Category.comp_id]
+        congr 2
+        dsimp
+        rw [← enrichedOrdinaryCategorySelf_eHomWhiskerRight,
+          ← enrichedOrdinaryCategorySelf_eHomWhiskerLeft]
+        let α : Under.mk (𝟙 j) ⟶ (Under.map φ).obj (Under.mk (𝟙 j')) := Under.homMk φ
+        exact (enrichedHom_condition C (Under.forget j ⋙ F₁) (Under.forget j ⋙ F₃) α).symm }
+  left_inv f := by aesop_cat
+  right_inv g := by
+    ext j
+    dsimp
+    ext k
+    rw [end_.lift_π, curry_uncurry, NatTrans.naturality_assoc]
+    dsimp
+    rw [end_.lift_π, enrichedOrdinaryCategorySelf_eHomWhiskerRight,
+      enrichedOrdinaryCategorySelf_eHomWhiskerLeft]
+    dsimp
+    rw [pre_id, NatTrans.id_app, Functor.map_id, Category.comp_id, Category.comp_id]
+    congr
+    dsimp [Under.map, Comma.mapLeft]
+    simp only [Category.comp_id]
+    rfl
```

Comment:
```suggestion
    simp? [enrichedOrdinaryCategorySelf_eHomWhiskerLeft, Under.map, Comma.mapLeft] says
      simp only [diagram_obj_obj, Functor.comp_obj, Under.forget_obj, enrichedCategorySelf_hom,
        curry_uncurry, NatTrans.naturality_assoc, functorEnrichedHom_obj, functorEnrichedHom_map,
        Under.map, Comma.mapLeft, Functor.const_obj_obj, Functor.id_obj, Discrete.natTrans_app,
        StructuredArrow.left_eq_id, end_.lift_π, Under.mk_right, Under.mk_hom, Iso.refl_inv,
        NatTrans.id_app, enrichedOrdinaryCategorySelf_eHomWhiskerRight, pre_id, Iso.refl_hom,
        enrichedOrdinaryCategorySelf_eHomWhiskerLeft, Functor.map_id, Category.comp_id]
    congr
    simp
```

Reply:
A slightly more automated version, if you'd like. But your version is maybe more readable, feel free
to ignore this suggestion.

Reply:
Thanks for the reviews! I have applied the modifications.